.circleci/config.yml

```diff
@@ -65,19 +65,41 @@ jobs:
           command: |
             shopt -s extglob
 
+            # Loop through unit tests
+            # Note: failed unit tests are skipped. Return the last failed unit
+            #   test exit code once the loop completes.
+
+            declare -i job_status=0
+
             for demo in !($IGNORED_DIRS); do
               if [ -d "$demo" ]; then
+                declare -i pid test_status=255
+
+                # run tests for this demo in bg
+                # subshell exits on first error
                 (
                   env="envs/$demo"
                   . "$env/bin/activate"
 
                   cd "$demo"
                   echo "$demo"
                   coverage run -m unittest discover
-                )
+                ) & pid=$!
+
+                # wait for tests to finish, collecting the exit status
+                # of the first failing command in the subshell
+                while (( test_status > 128 )); do
+                    wait $pid && test_status=0 || test_status=$?
+                done
+
+                # accumulate test status
+                # (don't fail for zero results in bash -e mode)
+                (( job_status |= test_status )) || true
```

Comment:
Previously, the unix tests were passing, but the linux tests were failing. You made a change with the `|| true` and now linux tests are passing again.

I see your comment, but I still don't understand why it works.

In my head:
- you're updating the `job_status` with the `test_status`
- when `job_status` is 0 (i.e. boolean false), run the return `true` in the list?

Questions:
- what is happening in this line of code?
- why did the previous code work for unix but not linux?

Reply:
Arithmetic expressions in POSIX have a quirk. Actually, a necessity for `if`/conditional execution to work as expected.

For example, to make these examples work:
```
if (( a == 2 )); then cmd; fi
(( a == 2 )) && cmd
```
exit status of `(( a == 2 ))` has to be "truthy" if `a` equals `2`. In Bash that means the exit status has to be `0` for "truthy" expressions. (And non-zero for "falsy".)

The expression in question, `job_status |= test_status`, will be (arithmetically) `0` -- "falsy". So the exit status of this command will be non-zero. Which, in turn, will terminate the Bash when run in `-e` mode.

Check [`((..))` section of Compound Commands](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Compound-Commands) in Bash manual.

Re: Unix/Linux, not sure what happened there. It shouldn't have worked on OSX. And when I re-run the tests today, they failed there too.